File: examples/VectorAddition/client/java/Dynamic/VectorAdditionClient.java
import com.maxeler.VectorAddition.VectorAdditionService;

import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/** VectorAddition Dynamic example. */
public final class VectorAdditionClient {

  /** Minimal random number. */
  private static final int MIN_RAND_NUM = 0;

  /** Maximal random number. */
  private static final int MAX_RAND_NUM = 1000;

  /** Size of int in bytes. */
  private static final int SIZE_OF_INT_IN_BYTES = 4;

  /** Size of int in bits. */
  private static final int SIZE_OF_INT = 32;

  /** Number of nano seconds in one second. */
  private static final int NUM_OF_NANO_SECONDS = 1000000000;

  /** Server port. */
  private static final int PORT = 9090;

  /** Utility classes should not have a public or default constructor. */
  private VectorAdditionClient() { }

  /**
   * Generates random data.
   *
   * @param size    Size
   *
   * @return        Random data
   */
  public static List<Integer> randomData(final int size) {
    List<Integer> randomData = new ArrayList<Integer>();

    Random rn = new Random();
    final int range = MAX_RAND_NUM - MIN_RAND_NUM + 1;

    for (int i = 0; i < size; i++) {
      randomData.add(rn.nextInt(range) + MIN_RAND_NUM);
    }

    return randomData;
  }

  /**
   * Checks if vectorAdditionDfe and vectorAdditionCpu return the same value.
   *
   * @param dataOutDfe  Data output from DFE
   * @param dataOutCpu  Data output from CPU
   * @param size        Size
   *
   * @return status     Number of elements that doesn't match
   */
  public static int check(final List<Integer> dataOutDfe,
                           final List<Integer> dataOutCpu, final int size) {
    int status = 0;

    for (int i = 0; i < size; i++) {
      if (!dataOutDfe.get(i).equals(dataOutCpu.get(i))) {
        System.out.println(
            "Output data @ " + i + " = " + dataOutDfe.get(i)
            + " (expected " + dataOutCpu.get(i) + ")");

        status++;
      }
    }

    return status;
  }

  /**
   * VectorAddition on CPU.
   *
   * @param size          Size
   * @param firstVector   First vector
   * @param secondVector  Second vector
   * @param scalar        Scalar
   *
   * @return              Data output
   */
  public static List<Integer> vectorAdditionCpu(
      final int size, final List<Integer> firstVector,
      final List<Integer> secondVector, final int scalar) {
    List<Integer> dataOut = new ArrayList<Integer>();

    for (int i = 0; i < size; i++) {
      dataOut.add(firstVector.get(i) + secondVector.get(i) + scalar);
    }

    return dataOut;
  }

  /**
   * VectorAddition on DFE.
   *
   * @param size          Size
   * @param firstVector   First vector
   * @param secondVector  Second vector
   * @param scalar        Scalar
   *
   * @return              Data output
   */
  public static List<Integer> vectorAdditionDfe(
      final int size, final List<Integer> firstVector,
      final List<Integer> secondVector, final int scalar) {
    List<Integer> dataOut = new ArrayList<Integer>();
    double startTime = System.nanoTime();
    DecimalFormat timeFormat = new DecimalFormat("#0.00000");

    // Make socket
    TTransport transport = new TSocket("localhost", PORT);

    // Wrap in a protocol
    TProtocol protocol = new TBinaryProtocol(transport);

    // Create a client to use the protocol encoder
    VectorAdditionService.Client client =
        new VectorAdditionService.Client(protocol);

    double estimatedTime = (System.nanoTime() - startTime)
                           / NUM_OF_NANO_SECONDS;
    System.out.println("Creating a client:\t\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

    try {
      // Connect!
      startTime = System.nanoTime();
      transport.open();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Opening connection:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Initialize maxfile
      startTime = System.nanoTime();
      final long maxfile = client.VectorAddition_init();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Initializing maxfile:\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Load DFE
      startTime = System.nanoTime();
      final long engine = client.max_load(maxfile, "*");
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Loading DFE:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Allocate and send input streams to server
      startTime = System.nanoTime();
      final long addressFirstVector = client.malloc_int32_t(size);
      client.send_data_int32_t(addressFirstVector, firstVector);
      final long addressSecondVector = client.malloc_int32_t(size);
      client.send_data_int32_t(addressSecondVector, secondVector);

      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Sending input data:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Allocate memory for output stream on server
      startTime = System.nanoTime();
      final long addressDataOut = client.malloc_int32_t(size);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Allocating memory for output stream on server:\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Action writeLMem
      startTime = System.nanoTime();
      final int sizeBytes = size * SIZE_OF_INT_IN_BYTES;

      long actions = client.max_actions_init(maxfile, "writeLMem");
      client.max_set_param_uint64t(actions, "address", 0);
      client.max_set_param_uint64t(actions, "nbytes", sizeBytes);
      client.max_queue_input(actions, "cpu_to_lmem",
                             addressFirstVector, sizeBytes);
      client.max_run(engine, actions);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Writing to LMem:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Action default
      startTime = System.nanoTime();

      actions = client.max_actions_init(maxfile, "default");
      client.max_set_param_uint64t(actions, "N", size);
      client.max_set_param_uint64t(actions, "A", scalar);
      client.max_queue_input(actions, "y", addressSecondVector, sizeBytes);
      client.max_queue_output(actions, "s", addressDataOut, sizeBytes);

      client.max_run(engine, actions);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Vector addition time:\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Unload DFE
      startTime = System.nanoTime();
      client.max_unload(engine);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Unloading DFE:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Get output stream from server
      startTime = System.nanoTime();
      dataOut = client.receive_data_int32_t(addressDataOut, size);
      System.out.println("Getting output stream:\t(size = "
                          + size * SIZE_OF_INT + " bit)\t"
                          + timeFormat.format(estimatedTime) + "s");

      // Free allocated memory for streams on server
      startTime = System.nanoTime();
      client.free(addressFirstVector);
      client.free(addressSecondVector);
      client.free(addressDataOut);
      client.free(actions);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Freeing allocated memory for streams on server:\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Free allocated maxfile data
      startTime = System.nanoTime();
      startTime = System.nanoTime();
      client.VectorAddition_free();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Freeing allocated maxfile data:\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Close!
      startTime = System.nanoTime();
      transport.close();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Closing connection:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

    } catch (TException x) {
      x.printStackTrace();
      System.exit(-1);
    }

    return dataOut;
  }

  /**
   * Calculates vectorAdditionDfe and vectorAdditionCpu and
   * checks if they return the same value.
   *
   * @param args Command line arguments
   */
  public static void main(final String[] args) {
    final int size = 384;
    final int scalar = 3;
    double startTime;
    double estimatedTime;
    DecimalFormat timeFormat = new DecimalFormat("#0.00000");

    // Generate input data
    startTime = System.nanoTime();
    final List<Integer> firstVector = randomData(size);
    final List<Integer> secondVector = randomData(size);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("Generating input data:\t\t\t\t"
                         + timeFormat.format(estimatedTime)
                         + "s");

    // DFE Output
    startTime = System.nanoTime();
    final List<Integer> dataOutDfe = vectorAdditionDfe(
        size, firstVector, secondVector, scalar);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("DFE vector addition total time:\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

    // CPU Output
    startTime = System.nanoTime();
    final List<Integer> dataOutCpu = vectorAdditionCpu(
        size, firstVector, secondVector, scalar);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("CPU vector addition total time:\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

    // Checking results
    startTime = System.nanoTime();
    int status = check(dataOutDfe, dataOutCpu, size);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("Checking results:\t\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");
    if (status == 0) {
      System.out.println("Test passed!");
    } else {
      System.out.println("Test failed " + status + " times!");
      System.exit(-1);
    }
  }
}

File: examples/VectorAddition/client/hs/Dynamic/VectorAdditionClient.hs
module Main where

import qualified VectorAdditionService_Client as Client
import VectorAddition_Types

import Thrift
import Thrift.Protocol.Binary
import Thrift.Server
import Thrift.Transport
import Thrift.Transport.Handle

import Control.Exception
import Data.Either
import Data.Int
import Data.List
import Data.Maybe
import Data.Time
import Data.Text.Lazy
import Data.Vector
import Network
import System.Exit
import System.Random
import Text.Printf

getRight           :: Either left right -> right
getRight (Right x) = x

randomlist :: Int -> StdGen -> [Int32]
randomlist  n = Data.List.take n . Data.List.unfoldr (Just . randomR (1, 1000))

vectorAdditionCPU :: [Int32] -> [Int32] -> Int32 -> [Int32]
vectorAdditionCPU [] [] s    = []
vectorAdditionCPU (x:xs) (y:ys) s = (x + y + s) : (vectorAdditionCPU (xs) (ys) s)

check :: [Int32] -> [Int32] -> Int -> Int -> [Int]
check [] [] start end = []
check (x:outDFE) (y:outCPU) start end
	| (x == y)		=  check outDFE outCPU (start+1) end
	| otherwise 	= (start) : check outDFE outCPU (start+1) end

printErrors :: [Int] -> [Int32] -> [Int32] -> String -> String
printErrors [] xs ys output                 = output
printErrors (i:is) xs ys output     = printErrors is xs ys (output Data.List.++ "Output data @ " Data.List.++ (show i) Data.List.++ " = " Data.List.++ (show (xs!!i))  Data.List.++ " (expected " Data.List.++ (show (ys!!i))  Data.List.++ ")\n")

main = do
    startTime <- getCurrentTime
    startDFETime <- getCurrentTime
    
    -- Make socket
    transport <- hOpen ("localhost", PortNumber 9090)

    -- Wrap in a protocol
    let protocol = BinaryProtocol transport

    -- Create a client to use the protocol encoder
    let client = (protocol, protocol)
    stopTime <- getCurrentTime
    putStrLn ("Creating a client and opening connection:\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Generate input data
    startTime <- getCurrentTime
    let size = 384
    let sizeBytes = size * 4
    let scalar = 3
    seed  <- newStdGen
    let x = randomlist size seed
    seed  <- newStdGen
    let y = randomlist size seed
    stopTime <- getCurrentTime
    putStrLn ("Generating input data:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Initialize maxfile
    startTime <- getCurrentTime
    e <- try (Client.vectorAddition_init client) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let maxfile = getRight e
    stopTime <- getCurrentTime
    putStrLn ("Initializing maxfile:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Load DFE
    startTime <- getCurrentTime
    e <- try (Client.max_load client maxfile (pack "*")) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let engine = getRight e
    stopTime <- getCurrentTime
    putStrLn ("Loading DFE:\t\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Allocate and send input streams to server
    startTime <- getCurrentTime

    e <- try (Client.malloc_int32_t client (fromIntegral size)) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_x = getRight e

    e <- try (Client.send_data_int32_t client address_x (fromList x)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.malloc_int32_t client (fromIntegral size)) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_y = getRight e

    e <- try (Client.send_data_int32_t client address_y (fromList y)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
        
    stopTime <- getCurrentTime
    putStrLn ("Sending input data:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Allocate memory for output stream on server
    startTime <- getCurrentTime

    e <- try (Client.malloc_int32_t client (fromIntegral size)) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_dataOut = getRight e

    stopTime <- getCurrentTime
    putStrLn ("Allocating memory for output stream on server:\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Writing to LMem
    startTime <- getCurrentTime

    e <- try (Client.max_actions_init client maxfile (pack "writeLMem")) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let actions = getRight e

    e <- try (Client.max_set_param_uint64t client actions (pack "address") (fromIntegral 0)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.max_set_param_uint64t client actions (pack "nbytes") (fromIntegral sizeBytes)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.max_queue_input client actions (pack "cpu_to_lmem") address_x (fromIntegral sizeBytes)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.max_run client engine actions) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    stopTime <- getCurrentTime
    putStrLn ("Writing to LMem:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Action default
    startTime <- getCurrentTime
    e <- try (Client.max_actions_init client maxfile (pack "default")) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let actions = getRight e

    e <- try (Client.max_set_param_uint64t client actions (pack "N") (fromIntegral size)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.max_set_param_uint64t client actions (pack "A") (fromIntegral scalar)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.max_queue_input client actions (pack "y") address_y (fromIntegral sizeBytes)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.max_queue_output client actions (pack "s") address_dataOut (fromIntegral sizeBytes)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.max_run client engine actions) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
 
    stopTime <- getCurrentTime
    putStrLn ("Vector addition time:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Unload DFE
    startTime <- getCurrentTime
    e <- try (Client.max_unload client engine) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    stopTime <- getCurrentTime
    putStrLn ("Unloading DFE:\t\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Get output stream from server
    startTime <- getCurrentTime
    dataOutDFE <- Client.receive_data_int32_t client address_dataOut (fromIntegral size)
    stopTime <- getCurrentTime
    putStrLn ("Getting output stream:\t(size = " Data.List.++ (show (size * 32)) Data.List.++ " bit)\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Free allocated memory for streams on server
    startTime <- getCurrentTime
    e <- try (Client.free client address_x) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    startTime <- getCurrentTime
    e <- try (Client.free client address_y) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.free client address_dataOut) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    
    stopTime <- getCurrentTime
    putStrLn ("Freeing allocated memory for streams on server:\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Free allocated maxfile data
    startTime <- getCurrentTime
    e <- try (Client.vectorAddition_free client) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    stopTime <- getCurrentTime
    putStrLn ("Freeing allocated maxfile data:\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Close!
    startTime <- getCurrentTime
    tClose transport
    stopTime <- getCurrentTime
    putStrLn ("Closing connection:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    stopTime <- getCurrentTime
    putStrLn ("DFE vector addition total time:\t\t\t" Data.List.++ (show (diffUTCTime stopTime startDFETime)))

    -- CPU Output
    startTime <- getCurrentTime
    let dataOutCPU = vectorAdditionCPU x y scalar
    stopTime <- getCurrentTime
    putStrLn ("CPU vector addition total time::\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Checking results
    startTime <- getCurrentTime
    let errors = check (toList dataOutDFE) dataOutCPU 0 size
    stopTime <- getCurrentTime
    putStrLn ("Checking results:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    if ((Data.List.length errors)== 0)
	then putStrLn ("Test successful!")
        else do putStr (printErrors errors (toList dataOutDFE) dataOutCPU [])
                putStrLn ("Test failed "  Data.List.++ show (Data.List.length errors)  Data.List.++ " times!")
                exitWith $ ExitFailure (-1)


